packages/block-editor/src/components/block-list/use-block-props/use-block-refs.js

```diff
@@ -29,57 +73,50 @@ export function useBlockRefProvider( clientId ) {
 	const { refs, callbacks } = useContext( BlockRefs );
 	const ref = useRef();
 	useLayoutEffect( () => {
-		refs.set( ref, clientId );
-		return () => {
-			refs.delete( ref );
-		};
-	}, [ clientId ] );
-	return useRefEffect(
+		addToMap( refs, clientId, ref );
+		return () => deleteFromMap( refs, clientId, ref );
+	}, [ refs, clientId ] );
+
+	return useCallback(
 		( element ) => {
+			if ( ! element ) {
+				return;
+			}
+
 			// Update the ref in the provider.
 			ref.current = element;
-			// Call any update functions.
-			callbacks.forEach( ( id, setElement ) => {
-				if ( clientId === id ) {
-					setElement( element );
-				}
-			} );
```

Comment:
Yes, here there should be a cleanup function that calls the callbacks with null.

Reply:
But previously, when the component's DOM changes, this was run again

Reply:
> there should be a cleanup function that calls the callbacks with null

This `useRefEffect` call didn't have a cleanup function. It did something only when the element was not `null`. It properly called `setElement` both with the initial element and also with the updated one, but `setElement( null )` was never called: the last element stayed in the local state forever, even when it was already removed from DOM.

Reply:
Yes, that's why I said "there should be a cleanup function" 🙂